### PR TITLE
vq diffusion classifier free sampling

### PR DESCRIPTION
**Pull request by williamberman**
Adds classifier free sampling to VQ diffusion. This results in significantly better image quality.

The pipeline now has a default guidance_scale of 5.0

Additionally, the ithq dataset uses a learned parameter for the classifier free embeddings. We modify the convert script to add this parameter to the ported model. **Weights will have to be reuploaded**

Prompts: "teddy bear playing in the pool" and "horse"

### Diffusers VQ diffusion with classifier free sampling
classifier_free_sampling

### Diffusers VQ diffusion without classifier free sampling
no_classifier_free_sampling

### Original VQ diffusion implementation with classifier free sampling
orig_classifier_free_sampling

### Original VQ diffusion implementation without classifier free sampling
orig_no_classifier_free_sampling

### CODE REVIEW

File: src/diffusers/models/embeddings.py [hunk 1]
```diff
@@ -198,3 +202,21 @@ def forward(self, index):
...
+
+        if self.learnable:
+            assert hidden_size is not None, "learnable=True requires `hidden_size` to be set"
+            assert length is not None, "learnable=True requires `length` to be set"
+
+            embeddings = torch.zeros(length, hidden_size)
+        else:
+            embeddings = None
+
+        self.embeddings = nn.Parameter(embeddings)
```
Comment (williamberman) [inline src/diffusers/models/embeddings.py:222]: Not sure if this is the prefferred way to add the learned embeddings to the pipeline. An alternative might be to add the additional vector to the scheduler instead

File: tests/pipelines/vq_diffusion/test_vq_diffusion.py [hunk 2]
```diff
@@ -173,3 +234,27 @@ def test_vq_diffusion(self):
 
         assert image.shape == (256, 256, 3)
         assert np.abs(expected_image - image).max() < 1e-2
+
+    def test_vq_diffusion_classifier_free_sampling(self):
+        expected_image = load_image(
+            "https://huggingface.co/datasets/williamberman/misc/resolve/main"
+            "/vq_diffusion/teddy_bear_pool_classifier_free_sampling.png"
```
Comment (williamberman) [inline tests/pipelines/vq_diffusion/test_vq_diffusion.py:241]: Should be moved to the huggingface testing dataset. FWIW you might have to also regenerate the image because I get a different image on `VQDiffusionPipelineIntegrationtests#test_vq_diffusion`.

File: src/diffusers/models/embeddings.py [hunk 3]
```diff
@@ -198,3 +202,25 @@ def forward(self, index):
...
```
Comment (williamberman) [inline src/diffusers/models/embeddings.py:226]: Not sure if this is the preferred way to add the learned embeddings to the pipeline. An alternative might be to add the additional vector to the scheduler instead

File: tests/pipelines/vq_diffusion/test_vq_diffusion.py (hunk 2 above)
Comment (patrickvonplaten) [reply tests/pipelines/vq_diffusion/test_vq_diffusion.py:241]: Very cool :-) I'll move it!

File: src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py [hunk 4]
```diff
@@ -64,6 +65,7 @@ def __init__(
         tokenizer: CLIPTokenizer,
         transformer: Transformer2DModel,
         scheduler: VQDiffusionScheduler,
+        learned_classifier_free_sampling_embeddings: LearnedClassifierFreeSamplingEmbeddings,
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py:68]: That's definitely the right way to do it - it's quite specific to vq-diffusion IMO though, so will move it here :-)

File: src/diffusers/models/embeddings.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/models/embeddings.py]: It's very model specific, so moving it to the pipeline here directly :-) 
Think that's a bit cleaner! The model works much better now though - thanks!
